File: LeetCode/Linked List/2130. Maximum Twin Sum of a Linked List.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        # Without using extra stack
        slow = head
        fast = head
        prev = None

        while fast and fast.next:
            fast = fast.next.next

            next_node = slow.next
            slow.next = prev
            prev = slow
            slow = next_node
        
        firstHalf = prev
        secondHalf = slow
        maxPair = 0

        while firstHalf and secondHalf:
            maxPair = max(maxPair, firstHalf.val + secondHalf.val)
            firstHalf = firstHalf.next
            secondHalf = secondHalf.next
        
        return maxPair

File: LeetCode/Linked List/2130. Maximum Twin Sum of a Linked List.js
//  Definition for singly-linked list.
function ListNode(val, next) {
  this.val = val === undefined ? 0 : val
  this.next = next === undefined ? null : next
}
/**
 * @param {ListNode} head
 * @return {number}
 */
var pairSum = function (head) {
  const nodeVals = []
  while (head) {
    nodeVals.push(head.val)
    head = head.next
  }
  let maxPair = 0
  const n = nodeVals.length
  for (let i = 0; i <= Math.floor(n / 2) - 1; i++) {
    maxPair = Math.max(maxPair, nodeVals[i] + nodeVals[n - i - 1])
  }
  return maxPair
}
